cmd/gateway/s3x/gateway-s3x-multipart.go

```diff
@@ -125,8 +133,67 @@ func (x *xObjects) CompleteMultipartUpload(
 ) (oi minio.ObjectInfo, e error) {
 	err := x.ledgerStore.AssertBucketExits(bucket)
 	if err != nil {
-		return oi, x.toMinioErr(err, bucket, "", "")
+		return oi, x.toMinioErr(err, bucket, object, uploadID)
 	}
-	return oi, errors.New("not finished yet")
-
+	m, unlock, err := x.ledgerStore.GetObjectDetails(uploadID)
+	if err != nil {
+		return oi, x.toMinioErr(err, bucket, object, uploadID)
+	}
+	defer unlock()
+	totalSize := uint64(0)
+	links := make([]*ipld.Link, 0, len(uploadedParts))
+	blocks := make([]uint64, 0, len(uploadedParts))
+	for _, p := range uploadedParts {
+		number := int64(p.PartNumber)
+		pi, ok := m.ObjectParts[number]
+		if !ok {
+			return oi, x.toMinioErr(fmt.Errorf("PartNumber %v not found", number), bucket, object, uploadID)
+		}
+		if pi.ActualSize <= 0 {
+			return oi, x.toMinioErr(fmt.Errorf("PartNumber %v reported ActualSize as %v", number, pi.ActualSize), bucket, object, uploadID)
+		}
+		cid, err := cid.Decode(pi.DataHash)
+		if err != nil {
+			return oi, x.toMinioErr(fmt.Errorf("PartNumber %v hash is not cid, %v", number, err), bucket, object, uploadID)
+		}
+		size := uint64(pi.ActualSize)
+		totalSize += size
+		links = append(links, &ipld.Link{
+			Size: size,
+			Cid:  cid,
+		})
+		blocks = append(blocks, size)
+	}
+	protoNode := &merkledag.ProtoNode{}
+	protoNode.SetCidBuilder(merkledag.V1CidPrefix())
+	protoNode.SetLinks(links)
+	data, err := proto.Marshal(&unixfs_pb.Data{
+		Type:       unixfs_pb.Data_File.Enum(),
+		Filesize:   &totalSize,
+		Blocksizes: blocks,
+	})
+	if err != nil {
+		return oi, x.toMinioErr(err, bucket, object, uploadID)
+	}
+	protoNode.SetData(data)
+	dataHash, err := ipfsSaveProtoNode(ctx, x.dagClient, protoNode)
```

Comment:
For the actual object data we need to save it as unixfs, otherwise if it is larger than 4MB we won't be able to send the gRPC request as 4MB is larger than the default gRPC message size. Additionally if the DAG object is larger than 4MB, we won't be able to send it through the network.

Reply:
The largest part number of a multipart upload is 10,000. So that puts an upper bound on the number of links and size of protoNode.

Reply:
If someone uploads a part that is larger than 4MB then wont this cause an error?

Reply:
Each part is uploaded using the file upload API

Reply:
Oh I see I missed that, good work.